---
title: "Brian Haas, Class Project MA681 2017"
author: "Brian Haas"
date: "12/14/2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(ggplot2)
library(dplyr)
```

```{r}
    rand_GM_vals_list <<- list()
    rand_H1_vals_list <<- list()
    
    perm_vals_mean_list <<- list()
    perm_vals_median_list <<- list()
    
    tested_actual_GM_vals <<- list()
    tested_actual_H1_vals <<- list()
```

```{r}
data = read.table("data/GM12892_H1hESC.counts.matrix", header=T, row.names=1)

# convert to CPMs
cs = colSums(data)
data = t(apply(data, 1, function(x) { x / cs * 1e6 }))

data = round(data) # integer counts

GM = data[,grep("^GM", colnames(data))]
H1 = data[,grep("^H1", colnames(data))]



```

Get variance calculations
```{r}
varGM = apply(GM, 1, var)
meanGM = rowMeans(GM)

varH1 = apply(H1, 1, var)
meanH1 = rowMeans(H1)

meanVarInfo = data.frame(meanGM=meanGM, varGM=varGM, meanH1=meanH1, varH1=varH1)
```

Show variance for each, compare to Poisson.

```{r}
plot(x = meanVarInfo$meanGM, y=meanVarInfo$varGM, log='xy', pch='.', xlab='mean(expression)', ylab='variance')
points(x=meanVarInfo$meanH1, y=meanVarInfo$varH1, col='blue', pch='.')
abline(a=0,b=1, col='red')
legend('topleft', legend=c('GM', 'H1'), fill=c('black', 'blue'))
```

```{r}
logfcs = log( (meanGM+1) / (meanH1 + 1) )
plot(density(logfcs), main='density log fold changes')
median(logfcs)
abline(v=0, col='red')
```

Identify differentially expressed transcripts

```{r}

perm_vals_stat = function(listA, listB, stat_func) {
 
    all_vals = c(listA, listB)
    perm_all_vals = sample(all_vals)
    new_listA = perm_all_vals[1:length(listA)]
    new_listB = perm_all_vals[(length(listA)+1):length(all_vals)]
    
    delta = stat_func(new_listA) - stat_func(new_listB)
    
    return(delta)
}
 
```


## Simple permutations to find DE genes:

```{r}
set.seed(1234)
genes = rownames(GM)
num_reps = 100
raw_perm_vals_mean = list()
#genes = names(logfcs)[which(abs(logfcs) > 2)]
fold_changes_df = data.frame(gene=NULL, logFC = NULL, p_val_for_mean=NULL, p_val_for_median=NULL)
counter = 0

GM_gene_vals = list()
H1_gene_vals = list()

for (gene in genes) {
  counter = counter + 1
  if (counter %% 1000 == 0) {
    message("at count: ", counter)
  }
  #if (counter > 1000) { break; }
  GM_vals = GM[gene,]
  H1_vals = H1[gene,]
  
  GM_gene_vals[[gene]] = GM_vals
  H1_gene_vals[[gene]] = H1_vals
  
  mean_GM = mean(GM_vals)
  mean_H1 = mean(H1_vals)
  
  logFC = log( (mean_GM+1) / (mean_H1+1) )
  
  diff_means = mean_GM - mean_H1
  
  perms_mean = replicate(n = num_reps, expr = perm_vals_stat(GM_vals, H1_vals, mean))
  
  raw_perm_vals_mean[[gene]] = perms_mean
  
  perms_median = replicate(n = num_reps, expr = perm_vals_stat(GM_vals, H1_vals, median))
  
  p_val_for_mean = sum( abs(perms_mean) >= abs(diff_means) ) / num_reps
  
  p_val_for_median = sum( abs(perms_median) >= abs(diff_means) ) / num_reps
  
  fold_changes_df = rbind(fold_changes_df, c(gene, logFC, p_val_for_mean, p_val_for_median))
  
}
colnames(fold_changes_df) = c('gene', 'logFC', 'p_val_for_mean', 'p_val_for_median')
write.table(fold_changes_df, file="simple_perm.fold_changes.dat", quote=F, sep='\t')
```


```{r}
system("./scripts/assign_prediction_status.pl simple_perm.fold_changes.dat 1 3 > simple_perm.fold_changes.dat.byMean.roc")
system("./scripts/assign_prediction_status.pl simple_perm.fold_changes.dat 1 4 > simple_perm.fold_changes.dat.byMedian.roc")
simple_perm_data_byMean = read.table("simple_perm.fold_changes.dat.byMean.roc")
plot(simple_perm_data_byMean$FP_rate, simple_perm_data_byMean$TP_rate,
     xlab='False Positive Rate', ylab='True Positive Rate', col='gray', t='l', lwd=3)
simple_perm_data_byMedian = read.table("simple_perm.fold_changes.dat.byMedian.roc")
lines(simple_perm_data_byMedian$FP_rate, simple_perm_data_byMedian$TP_rate, col='green', lwd=3)
legend(0.6, 0.4, legend=c('by mean', 'by median'), fill=c('gray', 'green'), box.lty=0)
```

```{r}

smoothScatter(fold_changes_df$logFC, -1*log(as.numeric(fold_changes_df$p_val_for_mean)), pch='.', xlab='log(fold change)', ylab='-1 * log(p)')

```






```{r}
save.image("session.RData")
```



```{r}
plot_gene_expr = function(gene) {
  GM_vals = tested_actual_GM_vals[[gene]]
  H1_vals = tested_actual_H1_vals[[gene]]
  
  GM_rand_vals = rand_GM_vals_list[[gene]]
  H1_rand_vals = rand_H1_vals_list[[gene]]
  
  message("mean GM_vals: ", mean(GM_vals), ", var GM_vals: ", var(GM_vals))
  message("mean GM_rand_vals: ", mean(GM_rand_vals), ", var GM_rand_vals: ", var(GM_rand_vals))
  
  message("mean H1_vals: ", mean(H1_vals), ", var H1_vals: ", var(H1_vals))
  message("mean H1_rand_vals: ", mean(H1_rand_vals), ", var H1_rand_vals: ", var(H1_rand_vals))
 
  message("Delta mean: ", abs(mean(GM_vals) - mean(H1_vals)))
  
  data = data.frame(type='GM_vals', pts = GM_vals)
  data = rbind(data, data.frame(type='GM_rand', pts = GM_rand_vals))
  
  data = rbind(data, data.frame(type='H1_vals', pts = H1_vals))
  data = rbind(data, data.frame(type='H1_rand', pts = H1_rand_vals))

  data %>% ggplot(aes(x=type, y=pts, color=type)) + geom_jitter(width=0.2)
  
}
  
```

```{r}
plot_dist_of_stat_differences = function(gene) {
  
    GM_vals = tested_actual_GM_vals[[gene]]
    H1_vals = tested_actual_H1_vals[[gene]]
  
    delta_mean = abs(mean(GM_vals) - mean(H1_vals))
    delta_median = abs(median(GM_vals) - median(H1_vals))
    
    raw_perm_means = abs(raw_perm_vals_mean[[gene]])
    
    perm_means = abs(perm_vals_mean_list[[gene]])
    perm_medians = abs(perm_vals_median_list[[gene]])
    
    df=data.frame(type='sim data', vals=perm_means)
    #df = rbind(df, data.frame(type='medians', vals=perm_medians))
    df = rbind(df, data.frame(type='orig data', vals=raw_perm_means))
    
    df %>% ggplot(aes(x=vals, color=type)) + geom_density() + 
      geom_vline(xintercept = delta_mean, color='black') 
    #+
      #geom_vline(xintercept = delta_median, color='green') 
    
} 
  
```


```{r}
#plot_dist_of_stat_differences('A1BG')
```



## Try using confidence in variance level to reduce false positives

```{r}

run_parameteric_permutations = function(numGM, numH1, 
                                        output_table_filename_prefix,
                                        use_log=F,
                                        add_noise=F,

                                        num_reps=50,
                                        var_weight = 1,
                                        
                                        varGM_func = NULL,
                                        varH1_func = NULL
                                        
                                        ) {

  
    fold_changes_df_results = data.frame(gene=NULL, logFC = NULL, 
                                 pvalue_for_mean=NULL, pvalue_for_median=NULL)
    
    
    rand_GM_vals_list <<- list()
    rand_H1_vals_list <<- list()
    
    perm_vals_mean_list <<- list()
    perm_vals_median_list <<- list()
    
    tested_actual_GM_vals <<- list()
    tested_actual_H1_vals <<- list()
    
    set.seed = 1234
    
    counter = 0
    genes = rownames(GM)
    #genes = c('A1BG')
    for (gene in genes) {
      counter = counter + 1
      if (counter %% 1000 == 0) {
        message("at count: ", counter)
      }
      #if (counter > 1000) { break}
      set.seed = 1234
      GM_vals = GM[gene,]
      H1_vals = H1[gene,]
      
      # add a little random noise
      if(add_noise) {
        GM_vals = GM_vals + rpois(n=length(GM_vals), lambda=1)
        H1_vals = H1_vals + rpois(n=length(H1_vals), lambda=1)
      }
      tested_actual_GM_vals[[gene]] <<- GM_vals
      tested_actual_H1_vals[[gene]] <<- H1_vals
      
      diff_means = mean(GM_vals) - mean(H1_vals)
      diff_medians = median(GM_vals) - median(H1_vals)
      
      # adjust to log space for variance estimates
      
      if (use_log) {
        GM_vals = log(GM_vals+1) # ensure non-negative
        H1_vals = log(H1_vals+1)
      }
      #GM_var_conf_level = quantile(x = replicate(10, var(sample(GM_vals, replace = T))), probs = 0.99)
      GM_var_conf_level = var(GM_vals) * var_weight
      gene_varGM = GM_var_conf_level
      
      if (! is.null(varGM_func)) {
        gene_varGM = varGM_func(gene)
      }
      
      gene_meanGM = mean(GM_vals)
      gene_medianGM = median(GM_vals)
      
      #H1_var_conf_level = quantile(x = replicate(10, var(sample(H1_vals, replace = T))), probs = 0.99)
      H1_var_conf_level = var(H1_vals) * var_weight
      gene_varH1 = H1_var_conf_level
      
      if (! is.null(varH1_func)) {
        gene_varH1 = varH1_func(gene)
      }
      
      gene_meanH1 = mean(H1_vals)
      gene_medianH1 = median(H1_vals)
      
      rnorm_GM = abs(rnorm(n = numGM, mean=gene_meanGM, sd=sqrt(gene_varGM)))
      #rnorm_GM = abs(rnorm(n = sample_size, mean=gene_meanGM, sd=sqrt(gene_varGM)))
      
      if (use_log) {
        rnorm_GM = exp(rnorm_GM) - 1  # convert back to non-log space
      }
      rand_GM_vals_list[[gene]] <<- rnorm_GM
      
      
      rnorm_H1 = abs(rnorm(n = numH1, mean=gene_meanH1, sd=sqrt(gene_varH1)))
      #rnorm_H1 = abs(rnorm(n = sample_size, mean=gene_meanH1, sd=sqrt(gene_varH1)))
      
      if (use_log) {
        rnorm_H1 = exp(rnorm_H1) - 1  # convert back to non-log space
      }
      rand_H1_vals_list[[gene]] <<- rnorm_H1
      
      logFC = log( (gene_meanGM+1) / (gene_meanH1+1) )
      
      perm_vals_mean =  replicate(n = num_reps, expr = perm_vals_stat(rnorm_GM, rnorm_H1, mean))
      perm_vals_mean_list[[gene]] <<- perm_vals_mean
      
      perm_vals_median =  replicate(n = num_reps, expr = perm_vals_stat(rnorm_GM, rnorm_H1, median))
      perm_vals_median_list[[gene]] <<- perm_vals_median
      
      p_val_for_mean = sum( abs(perm_vals_mean) >= abs(diff_means) ) / num_reps
      
      p_val_for_median = sum( abs(perm_vals_median) >= abs(diff_medians)) / num_reps
      
      fold_changes_df_results = rbind(fold_changes_df_results, 
                                          c(gene, logFC, 
                                            p_val_for_mean, p_val_for_median))
      
      
    }
    
    colnames(fold_changes_df_results) = c('gene', 'logFC', 'p_val_for_mean', 'p_val_for_median')
    
    write.table(fold_changes_df_results, file=paste0(output_table_filename_prefix,".dat"), 
                quote=F, sep='\t')
    
    cmd = paste0("./scripts/assign_prediction_status.pl ", output_table_filename_prefix,".dat ",
                " 1 3 > ", output_table_filename_prefix,".dat.byMean.roc")
    system(cmd)
    
    cmd = paste0("./scripts/assign_prediction_status.pl ", output_table_filename_prefix,".dat ",
                " 1 4 > ", output_table_filename_prefix,".dat.byMedian.roc")
    
    system(cmd)
   
     
  sim_data_byMean = read.table(paste0(output_table_filename_prefix,".dat.byMean.roc"))
  sim_data_byMedian = read.table(paste0(output_table_filename_prefix,".dat.byMedian.roc"))
  
  ret = list(byMean = sim_data_byMean,
             byMedian = sim_data_byMedian)
  
  return(ret)
  
}
```

```{r}

results_4_3 = run_parameteric_permutations(numGM = length(GM_vals),
                                       numH1 = length(H1_vals),
                                       output_table_filename_prefix = "sim_4_3")
```


```{r}

plot(simple_perm_data_byMean$FP_rate, 
     simple_perm_data_byMean$TP_rate, t='l', col='gray')
points(simple_perm_data_byMedian$FP_rate, 
       simple_perm_data_byMedian$TP_rate, t='l', col='green')



points(results_4_3[['byMean']]$FP_rate, results_4_3[['byMean']]$TP_rate, col='red', t='l')

points(results_4_3[['byMedian']]$FP_rate, results_4_3[['byMedian']]$TP_rate, col='pink', t='l')



```


```{r}

results_5_5 = run_parameteric_permutations(numGM = 5,
                                       numH1 = 5,
                                       output_table_filename_prefix = "sim_5_5")
```


```{r}

results_10_10 = run_parameteric_permutations(numGM = 10,
                                       numH1 = 10,
                                       output_table_filename_prefix = "sim_10_10")
```


```{r}

results_20_20 = run_parameteric_permutations(numGM = 20,
                                       numH1 = 20,
                                       output_table_filename_prefix = "sim_20_20")
```


```{r}

results_30_30 = run_parameteric_permutations(numGM = 30,
                                       numH1 = 30,
                                       output_table_filename_prefix = "sim_30_30")
```



```{r}

plot(simple_perm_data_byMean$FP_rate, simple_perm_data_byMean$TP_rate, t='l', col='gray', xlab='False positive rate', ylab='True positive rate')

points(results_4_3[['byMean']]$FP_rate, results_4_3[['byMean']]$TP_rate, col='red', t='l')

#points(results_5_5[['byMean']]$FP_rate, results_5_5[['byMean']]$TP_rate, col='green', t='l')

points(results_10_10[['byMean']]$FP_rate, results_10_10[['byMean']]$TP_rate, col='orange', t='l')

points(results_20_20[['byMean']]$FP_rate, results_20_20[['byMean']]$TP_rate, col='cyan', t='l')

points(results_30_30[['byMean']]$FP_rate, results_30_30[['byMean']]$TP_rate, col='magenta', t='l')

legend(0.4, 0.6, legend=c('orig data', 'sim=4x3', 'sim=10x10', 'sim=20x20', 'sim=30x30'),
       fill=c('gray', 'red', 'orange', 'cyan', 'magenta'), box.lty=0)

```

examine a false positive

```{r}
gene_name = 'ADAMTS12'
plot_dist_of_stat_differences(gene_name)
plot_gene_expr(gene_name)

```



## Using trended variance by leveraging linear modeling
  
GM linear modeling
  
```{r}
#plot(x = meanVarInfo$meanGM, y=meanVarInfo$varGM, log='xy', pch='.')
#points(x=meanVarInfo$meanH1, y=meanVarInfo$varH1, col='blue', pch='.')
#abline(a=0,b=1, col='red')

gm_mod = lm(log(meanVarInfo$varGM +1) ~ log(meanVarInfo$meanGM +1))
gm_poly_mod = lm(log(meanVarInfo$varGM +1) ~ log(meanVarInfo$meanGM +1) + poly(log(meanVarInfo$meanGM +1), 2))

gm_fitted_logvar = h1_poly_mod$fitted.values
names(gm_fitted_logvar) = rownames(meanVarInfo)

plot(x = meanVarInfo$meanGM, y=meanVarInfo$varGM, log='xy', pch='.',
     xlab='mean(expression)', ylab='variance', main='sample GM: modeling variance ~ mean')
points(x=meanVarInfo$meanGM, y=exp(predict(gm_mod, data=log(meanVarInfo$meanGM +1))), col='green')
points(x=meanVarInfo$meanGM, y=exp(predict(gm_poly_mod, data=log(meanVarInfo$meanGM +1))), col='orange')
AIC(gm_mod)
AIC(gm_poly_mod)

legend("topleft", legend=c('y~x', 'y~x + poly(x,2)'), fill=c('green', 'orange'), box.lty=0)

```
  
  
  
H1 linear modeling  
  
```{r}

h1_mod = lm(log(meanVarInfo$varH1 +1) ~ log(meanVarInfo$meanH1 +1))
h1_poly_mod = lm(log(meanVarInfo$varH1 +1) ~ log(meanVarInfo$meanH1 +1) + poly(log(meanVarInfo$meanH1 +1), 2))

plot(x = meanVarInfo$meanH1, y=meanVarInfo$varH1, log='xy', pch='.',
       xlab='mean(expression)', ylab='variance', 
     main='sample H1: modeling variance ~ mean')

points(x=meanVarInfo$meanH1, y=exp(predict(h1_mod, data=log(meanVarInfo$meanH1 +1))), col='green')
points(x=meanVarInfo$meanH1, y=exp(predict(h1_poly_mod, data=log(meanVarInfo$meanH1 +1))), col='orange')
AIC(h1_mod)
AIC(h1_poly_mod)

h1_fitted_logvar = h1_poly_mod$fitted.values
names(h1_fitted_logvar) = rownames(meanVarInfo)

legend("topleft", legend=c('y~x', 'y~x + poly(x,2)'), fill=c('green', 'orange'), box.lty=0)
```

  
## Use the linear modeling of the variance

Setting lower bound for variance based on Poisson

```{r}

varGM_func = function(gene) {
  return(max(
    mean(GM[gene,]), 
    exp(gm_fitted_logvar[gene])-1)
    )
}

varH1_func = function(gene) {
  return(max(
    mean(H1[gene,]),
    exp(h1_fitted_logvar[gene])-1)
  )
}

results_useLMtrend = run_parameteric_permutations(numGM = length(GM_vals),
                             numH1 = length(H1_vals),
                              output_table_filename_prefix = "sim_4_3_useLMtrend",
                             varGM_func = varGM_func,
                             varH1_func = varH1_func)

```

```{r}
plot(simple_perm_data_byMean$FP_rate, simple_perm_data_byMean$TP_rate, t='l', col='gray', xlab='False positive rate', ylab='True positive rate')

points(results_4_3[['byMean']]$FP_rate, results_4_3[['byMean']]$TP_rate, col='red', t='l')

points(results_useLMtrend[['byMean']]$FP_rate, results_useLMtrend[['byMean']]$TP_rate, col='brown', t='l')



```



Try using peak measurements, with trended variance as lower bound


```{r}


varGM_func = function(gene) {
  return(max(
    var(GM[gene,]), 
    mean(GM[gene,]),
    exp(gm_fitted_logvar[gene])-1)
    )
}

varH1_func = function(gene) {
  return(max(
    var(H1[gene,]),
    mean(H1[gene,]),
    exp(h1_fitted_logvar[gene])-1)
  )
}

results_useLMtrend_orMaxVar = run_parameteric_permutations(numGM = length(GM_vals),
                             numH1 = length(H1_vals),
                      output_table_filename_prefix = "sim_4_3_useLMtrend_orMaxVar",
                             varGM_func = varGM_func,
                             varH1_func = varH1_func)

```


```{r}

plot(simple_perm_data_byMean$FP_rate, simple_perm_data_byMean$TP_rate, t='l', col='gray', xlab='False positive rate', ylab='True positive rate', xlim=c(0,0.4), lwd=3)

points(results_4_3[['byMean']]$FP_rate, results_4_3[['byMean']]$TP_rate, col='red', t='l')

points(results_useLMtrend[['byMean']]$FP_rate, results_useLMtrend[['byMean']]$TP_rate, col='brown', t='l')

points(results_useLMtrend_orMaxVar[['byMean']]$FP_rate, results_useLMtrend_orMaxVar[['byMean']]$TP_rate, col='magenta', t='l')

legend(0.15, 0.6, legend=c('orig data', 'sim=4x3', 'trended_var=4x3', 'trended_var_or_max=4x3'), fill=c('gray', 'red', 'brown', 'magenta'), box.lty=0)

```


